File: Linked  List/2. Add Two Numbers/Solution.java
/*
 时间复杂度 O(n)
 空间复杂度 O(n) 

 模拟加法，因为是反向存储，最后反向输出，所以可以模拟，注意进位操作

*/


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int count = 0;
        ListNode head = new ListNode(0);
        ListNode p3 = head;
        while(l1 != null || l2 != null){
            if(l1 != null){
                count += l1.val;
                l1 = l1.next;
            }
            if(l2 != null){
                count += l2.val;
                l2 = l2.next;
            }
            p3.next = new ListNode(count%10);
            p3 = p3.next;
            count /= 10; 
        }
        if(count==1){
            p3.next = new ListNode(count);
            
        }
        
        return head.next;
    }
}

File: Linked  List/2. Add Two Numbers/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        count = 0
        head = ListNode(0)
        l3 = head
        while l1 or l2:
            if l1:
                count += l1.val
                l1 = l1.next
            if l2:
                count += l2.val
                l2 = l2.next
            l3.next = ListNode(count%10)
            l3 = l3.next
            count = count//10
        if count == 1:
            l3.next = ListNode(count)
        return head.next